src/posts/flox-smart/index.md

```diff
@@ -0,0 +1,215 @@
+---
+title: 'flox: GroupBy, now with smarts!'
+date: '2024-05-31'
+authors:
+  - name: Deepak Cherian
+    github: dcherian
+
+summary: 'flox adds heuristics for automatically choosing an appropriate reduction strategy with dask arrays!'
+---
+
+## TL;DR
+
+`flox>=0.9` now automatically optimizes GroupBy reductions with dask arrays to reduce memory usage and increase parallelism! Here I describe how. To opt in, simply set `method=None` if you were setting it explicitly previously.
+
+## What is flox?
+
+[`flox` implements](https://flox.readthedocs.io/) grouped reductions for chunked array types like [cubed](https://cubed-dev.github.io/cubed/) and [dask](https://docs.dask.org/en/stable/array.html) using tree reductions.
+Tree reductions ([example](https://people.csail.mit.edu/xchen/gpu-programming/Lecture11-reduction.pdf)) are a parallel-friendly way of computing common reduction operations like `sum`, `mean` etc.
+Briefly, one computes the reduction for a subset of the array N chunks at a time in parallel, then combines those results together again N chunks at a time, until we have the final result.
+
+Without flox, Xarray effectively shuffles — sorts the data to extract all values in a single group — and then runs the reduction group-by-group.
+Depending on data layout or "chunking" this shuffle can be quite expensive.
+Here's a schematic of an array with 5 chunks arranged vertically, and each chunk has 10 elements each of which are colored by group.
+![shuffle](https://flox.readthedocs.io/en/latest/_images/new-split-apply-combine-annotated.svg)
+With flox installed, Xarray instead uses its parallel-friendly tree reduction.
+In many cases, this is a massive improvement.
+Notice how much cleaner the graph is in this image:
+![map-reduce](https://flox.readthedocs.io/en/latest/_images/new-map-reduce-reindex-True-annotated.svg)
+See our [previous blog post](https://xarray.dev/blog/flox) or the [docs](https://flox.readthedocs.io/en/latest/implementation.html#method-map-reduce) for more.
```

Comment:
It's a little confusing that these two images are not for identical problems - the inputs are different.